### ISSUE TITLE
StackDump still fails with MemoryError on Posts.xml processed by stackdump-lowram

### ISSUE THREAD

**Issue by JGKle** [RubenNL/stackdump-lowram#4]
Running Posts.xml through the current master, then trying to build using the StackDump from https://github.com/RubenNL/stackdump, still seems to yield a MemoryError.  stackdump-lowram seemed to complete successfully (though it didn't show any output, as a previous version had), and I ended up with a new Posts.xml (though the original was removed, so I wasn't able to compare how much smaller it ended up). When building StackDump from the resulting posts.xml, it made it through Comments.xml & Users.xml, then died during posts.xml.  

Screenshot: https://lnk2.tk/snaps/2020-06-05_22.13.08.png

```
Traceback (most recent call last):
  File "./python/src/stackdump/commands/import_site.py", line 927, in <module>
    cmd_options.base_url, answer_yes=cmd_options.answer_yes)
  File "./python/src/stackdump/commands/import_site.py", line 885, in import_site
    xml.sax.parse(xml_path, handler)
  File "/usr/local/lib/python2.7/xml/sax/__init__.py", line 33, in parse
    parser.parse(source)
  File "/usr/local/lib/python2.7/xml/sax/expatreader.py", line 111, in parse
    xmlreader.IncrementalParser.parse(self, source)
  File "/usr/local/lib/python2.7/xml/sax/xmlreader.py", line 123, in parse
    self.feed(buffer)
  File "/usr/local/lib/python2.7/xml/sax/expatreader.py", line 220, in feed
    self._parser.Parse(data, isFinal)
  File "/usr/local/lib/python2.7/xml/sax/expatreader.py", line 339, in end_element
    self._cont_handler.endElement(name)
  File "./python/src/stackdump/commands/import_site.py", line 364, in endElement
    self.commit_finished_questions()
  File "./python/src/stackdump/commands/import_site.py", line 390, in commit_finished_questions
    self.commit_questions(questions_to_commit, commit=False)
  File "./python/src/stackdump/commands/import_site.py", line 548, in commit_questions
    self.solr.add(questions, commit=commit)
  File "/app/stackdump/python/packages/pysolr.py", line 794, in add
    return self._update(m, commit=commit, waitFlush=waitFlush, waitSearcher=waitSearcher)
  File "/app/stackdump/python/packages/pysolr.py", line 399, in _update
    return self._send_request('post', path, message, {'Content-type': 'text/xml; charset=utf-8'})
  File "/app/stackdump/python/packages/pysolr.py", line 291, in _send_request
    log_body = repr(body)
MemoryError

**Comment by JGKle** [comment]
Out of desperation to get this working, I went out & bought more physical RAM.  Still no go.  Using the Posts.xml produced by the current master & the version of stackdump from the RubenNL repo, it gets farther now, but always dies after a few hours of working on posts, with a message like:

```
...
[post]     Processed 2429000 rows.
[post]     Processed 2430000 rows.
Committing completed questions...
A Solr error occurred while committing questions -
Traceback (most recent call last):
  File "./python/src/stackdump/commands/import_site.py", line 548, in commit_questions
    self.solr.add(questions, commit=commit)
  File "/app/stackdump/python/packages/pysolr.py", line 794, in add
    return self._update(m, commit=commit, waitFlush=waitFlush, waitSearcher=waitSearcher)
  File "/app/stackdump/python/packages/pysolr.py", line 399, in _update
    return self._send_request('post', path, message, {'Content-type': 'text/xml; charset=utf-8'})
  File "/app/stackdump/python/packages/pysolr.py", line 323, in _send_request
    raise SolrError(error_message % params)
SolrError: Failed to connect to server at 'http://localhost:8983/solr/stackdump/update/?commit=false', are you sure that URL is correct? Checking it in a browser might help: HTTPConnectionPool(host='localhost', port=8983): Max retries exceeded with url: /solr/stackdump/update/?commit=false (Caused by <class 'socket.error'>: [Errno 99] Address not available)

Try committing the questions again? (y/n) Traceback (most recent call last):
  File "./python/src/stackdump/commands/import_site.py", line 927, in <module>
    cmd_options.base_url, answer_yes=cmd_options.answer_yes)
  File "./python/src/stackdump/commands/import_site.py", line 885, in import_site
    xml.sax.parse(xml_path, handler)
  File "/usr/local/lib/python2.7/xml/sax/__init__.py", line 33, in parse
    parser.parse(source)
  File "/usr/local/lib/python2.7/xml/sax/expatreader.py", line 111, in parse
    xmlreader.IncrementalParser.parse(self, source)
  File "/usr/local/lib/python2.7/xml/sax/xmlreader.py", line 123, in parse
    self.feed(buffer)
  File "/usr/local/lib/python2.7/xml/sax/expatreader.py", line 220, in feed
    self._parser.Parse(data, isFinal)
  File "/usr/local/lib/python2.7/xml/sax/expatreader.py", line 339, in end_element
    self._cont_handler.endElement(name)
  File "./python/src/stackdump/commands/import_site.py", line 364, in endElement
    self.commit_finished_questions()
  File "./python/src/stackdump/commands/import_site.py", line 390, in commit_finished_questions
    self.commit_questions(questions_to_commit, commit=False)
  File "./python/src/stackdump/commands/import_site.py", line 555, in commit_questions
    response = raw_input('Try committing the questions again? (y/n) ').lower()
EOFError: EOF when reading a line
```

**Comment by teoman002** [comment]
Could you post the last 10 lines of the post.xml
tail -n 10 path/to/file

**Comment by JGKle** [comment]
```
62111828  <row Id="62111840" PostTypeId="2" ParentId="62111828" CreationDate="2020-05-31T05:09:45.353" Score="0" Body="&lt;p&gt;&lt;code&gt;input&lt;/code&gt; means that it's waiting for you to write something.&#xA;Type something and press&quot;Enter&quot;&lt;/p&gt;&#xA;&#xA;&lt;p&gt;It isn't stopped.&lt;/p&gt;&#xA;" OwnerUserId="11915854" LastActivityDate="2020-05-31T05:09:45.353" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111828  <row Id="62111867" PostTypeId="2" ParentId="62111828" CreationDate="2020-05-31T05:14:48.220" Score="0" Body="&lt;p&gt;you need to type some thing after the third line like&#xA;$Mizu&quot;Mizu&quot;.&#xA;The Mizu shown on the run is just a reminder message to let know which input are you giving value.&lt;/p&gt;&#xA;" OwnerUserId="13651122" LastActivityDate="2020-05-31T05:14:48.220" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111828  <row Id="62111882" PostTypeId="2" ParentId="62111828" CreationDate="2020-05-31T05:17:19.107" Score="0" Body="&lt;p&gt;The programme is not stopped. It is waiting for your input.&#xA;In the input function you have to give some valut at the run time.&#xA;myName = input(&quot;Mizu&quot;). In this line you have entered your question. You have to give answer at run time&#xA;myName = input(&quot;what is your name&quot;)&lt;/p&gt;&#xA;" OwnerUserId="13631307" LastActivityDate="2020-05-31T05:17:19.107" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111847  <row Id="62111847" PostTypeId="1" CreationDate="2020-05-31T05:10:42.250" Score="0" ViewCount="8" Body="&lt;p&gt;Why does my code print: TypeError: argument of type 'int' is not iterable, whenever I run it? I have the following lists given below. What I am trying to do is to get the element/s in O then find its match in J. After finding the match, the index is obtained. Using this index, I will find its corresponding value in R.&lt;/p&gt;&#xA;&#xA;&lt;pre&gt;&lt;code&gt;    J = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]&#xA;    R = [[0, 0], [4, 3], [2, 0], [2, 1], [0, 3], [1, 1], [1, 2], [2, 1], [0, 3], [2, 2], [0, 0]]&#xA;    O = J[0]&#xA;&#xA;    R_demand = [d for a, d in zip(J, R) if a in O]&#xA;    print(R_demand)&#xA;&lt;/code&gt;&lt;/pre&gt;&#xA;&#xA;&lt;p&gt;For the code above, since O = 1 in J, I should get R_demand = [0, 0] as my answer. All values in lists are integers. &lt;/p&gt;&#xA;" OwnerUserId="13618572" LastActivityDate="2020-05-31T05:30:54.193" Title="TypeError: argument of type 'int' is not iterable in indexing" Tags="&lt;python-3.x&gt;&lt;zip&gt;&lt;typeerror&gt;" AnswerCount="1" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111847  <row Id="62111963" PostTypeId="2" ParentId="62111847" CreationDate="2020-05-31T05:30:54.193" Score="0" Body="&lt;p&gt;You are correct, &lt;code&gt;O = 1&lt;/code&gt;. Which is why the conditional part of your list comprehension &lt;code&gt;if a in O&lt;/code&gt; throws a TypeError. The &lt;code&gt;in&lt;/code&gt; operation needs an iterable to follow it, 1 is a constant (non-iterable), so &lt;code&gt;a in 1&lt;/code&gt; is invalid. If O will always be an integer, just change &lt;code&gt;in&lt;/code&gt; to &lt;code&gt;==&lt;/code&gt;, as in:&lt;/p&gt;&#xA;&#xA;&lt;pre&gt;&lt;code&gt;R_demand = [d for a, d in zip(J, R) if a == O]&#xA;&lt;/code&gt;&lt;/pre&gt;&#xA;&#xA;&lt;p&gt;Hope that helps, happy coding!&lt;/p&gt;&#xA;" OwnerUserId="6296908" LastActivityDate="2020-05-31T05:30:54.193" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111911  <row Id="62111911" PostTypeId="1" CreationDate="2020-05-31T05:22:16.967" Score="-1" ViewCount="4" Body="&lt;p&gt;I wish to use only the git tool and not any other tools to do this. I am looking for a shell script or even it is a inbuilt git sub command which can update all of my local branches by pulling changes from the origin master. So that I can simply run the script or the sub command to get this done.&#xA;Please help.&lt;/p&gt;&#xA;" OwnerUserId="10069193" LastActivityDate="2020-05-31T05:28:45.170" Title="how to update all my local git branches with a single script?" Tags="&lt;git&gt;&lt;shell&gt;&lt;git-bash&gt;" AnswerCount="1" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111911  <row Id="62111951" PostTypeId="2" ParentId="62111911" CreationDate="2020-05-31T05:28:45.170" Score="0" Body="&lt;p&gt;you can use &lt;code&gt;git pull --all&lt;/code&gt; to update all your local branches.&lt;/p&gt;&#xA;" OwnerUserId="10738278" LastActivityDate="2020-05-31T05:28:45.170" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111929  <row Id="62111929" PostTypeId="1" CreationDate="2020-05-31T05:25:14.593" Score="0" ViewCount="8" Body="&lt;p&gt;I am developing a reactjs application which uses unsplash API to update background images of the application. The result from the API is as below:&lt;/p&gt;&#xA;&#xA;&lt;pre&gt;&lt;code&gt;{total: 2330, total_pages: 78, results: Array(30)}&#xA;&lt;/code&gt;&lt;/pre&gt;&#xA;&#xA;&lt;p&gt;As the API gives back 30 images only per page, I have implemented the following logic to retrieve any random image from the list of 30 images:&lt;/p&gt;&#xA;&#xA;&lt;pre&gt;&lt;code&gt;        if (typeof response.data !== &quot;undefined&quot;) {&#xA;          const num = Math.floor(Math.random() * 30) + 1;&#xA;          const url = response.data.results[num].urls.regular;&#xA;          updateBackground(url);&#xA;        }&#xA;&lt;/code&gt;&lt;/pre&gt;&#xA;&#xA;&lt;p&gt;But the above logic is limited to picking up a random image from the first page only. How do I pick a random image from the total of 2330 images ? Is there a way to exclude total_pages from the result ?&lt;/p&gt;&#xA;&#xA;&lt;p&gt;Thanks&lt;/p&gt;&#xA;" OwnerUserId="1324023" LastActivityDate="2020-05-31T05:29:18.383" Title="pick a random image from unsplash API reactjs" Tags="&lt;javascript&gt;&lt;reactjs&gt;&lt;api&gt;&lt;random&gt;" AnswerCount="1" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
62111929  <row Id="62111956" PostTypeId="2" ParentId="62111929" CreationDate="2020-05-31T05:29:18.383" Score="0" Body="&lt;p&gt;Wouldn't the Random Photo API be the solution you need? &lt;a href=&quot;https://unsplash.com/documentation#get-a-random-photo&quot; rel=&quot;nofollow noreferrer&quot;&gt;https://unsplash.com/documentation#get-a-random-photo&lt;/a&gt;&lt;/p&gt;&#xA;&#xA;&lt;p&gt;Otherwise if that's not what you need, you could randomize the page you pick from in addition to the photo within that page, perhaps.&lt;/p&gt;&#xA;" OwnerUserId="1950503" LastActivityDate="2020-05-31T05:29:18.383" CommentCount="0" ContentLicense="CC BY-SA 4.0" />
</posts>
```

**Comment by JGKle** [comment]
Incidentally: I've retried it, completely from scratch, on four separate occasions.  It always runs happily for a while before getting to that `A Solr error occurred while committing questions`  Once it gets there, I can hit 'y' to retry as many times as I like - it never continues.  It doesn't always die at the same place - i.e. I just tried it again, this time it got to `Processed 2710000 rows` before it died.

**Comment by JGKle** [comment]
More info: what's happening is solr keeps dying.

When StackDump is asking `Try committing the questions again? (y/n)`, I've found that I can open a second prompt and restart it via `./stackdump/start_solr.sh`.  Then in the original prompt, I can answer 'y' and it will continue to process a few more rows.  However, almost immediately it dies again.  If I go back & forth between these 2 windows, restarting, answering 'yes,' restarting, answering 'yes,' and it does continue to make progress - but obviously this is agonizingly slow & would probably take about a week of continuous user intervention to actually fully complete the process.   So the question is: why does solr keep shutting down?  Note: it always seems to start happening between 2000000 & 3000000 rows.

Right before solr kills itself, the last few lines in that window are:

```
46470 [commitScheduler-6-thread-1] INFO  org.apache.solr.core.SolrCore  – newest commit generation = 149
46591 [commitScheduler-6-thread-1] INFO  org.apache.solr.search.SolrIndexSearcher  – Opening Searcher@fb831f5 realtime
47049 [commitScheduler-6-thread-1] INFO  org.apache.solr.update.UpdateHandler  – end_commit_flush
./stackdump/start_solr.sh: line 25:   268 Killed                  "$JAVA_CMD" -Xmx2048M -XX:MaxPermSize=512M -Djetty.host=127.0.0.1 -jar start.jar
```

**Edit:** I also tried restarting solr with larger memory values in both cases: -Xmx2400M and -XX:MaxPermSize=768M. Same behavior.
I also tried with smaller memory values in both places (1400M and 256M). Same behavior.

**Comment by JGKle** [comment]
Solved by setting up a baremetal Linux system & running it there.

Originally I had been running StackDump on Windows, but this caused far too many issues/conflicts due to StackDump's use of the now extremely outdated Python2, which causes all kinds of issues if installed alongside Python3 (on Win).

Thus I moved it to Docker, where it's been running well for at least 2-3 years.  The above errors were occurring when trying to import a new dump via a similar Docker container (using the RubenNL copy of StackDump). Per above, imports always failed.

When I installed a whole standalone Linux instance, the import was finally able to complete.  Then I took the finished /data dir & moved it back to Docker, so I could use it from both Linux & Windows.

Yeesh, I don't even want to think about how many days went down the drain working on this. So the lesson to anyone else who finds this is:

Don't import a StackOverflow data dump to a StackDump instance running in Docker.  Docker works fine for running it, but not for importing.
